lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -441,6 +1975,244 @@ bool StackAnalysis::runOnModule(Module &M) {
+                                         "retcall_hook",
+                                         M);
+  RetHookMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  RetHookMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  // Initialize marker to adjust `unexpectedpc` basic block for fake functions
+  FTy = llvm::FunctionType::get(llvm::Type::getVoidTy(M.getContext()), false);
+  auto
+    *UnexpectedPCMarker = Function::Create(FTy,
+                                           llvm::GlobalValue::ExternalLinkage,
+                                           "unexpectedpc_hook",
+                                           M);
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::ReadOnly);
+  UnexpectedPCMarker->addFnAttr(llvm::Attribute::NoUnwind);
+
+  // Queue to be populated with the CFEP
+  llvm::SmallVector<BasicBlock *, 8> Worklist;
+  llvm::SmallVector<BasicBlockNode *, 4> EntryBlocks;
+  UniquedQueue<BasicBlockNode *> CFEPQueue;
+  SmallCallGraph CG{};
+
+  using BasicBlockToNodeMapTy = llvm::DenseMap<BasicBlock *, BasicBlockNode *>;
+  BasicBlockToNodeMapTy BasicBlockNodeMap;
+
+  // Create over-approximated call graph
+  for (CFEP &Function : Functions) {
+    BasicBlockNode Node{ Function.Entry };
+    auto *GraphNode = CG.addNode(Node);
+    BasicBlockNodeMap[Function.Entry] = GraphNode;
+  }
+
+  for (CFEP &Function : Functions) {
+    llvm::SmallSet<BasicBlock *, 8> Visited;
+    auto *StartNode = BasicBlockNodeMap[Function.Entry];
+    Worklist.emplace_back(Function.Entry);
+
+    do {
+      auto *Current = Worklist.pop_back_val();
+      Visited.insert(Current);
+      for (auto *Successor : successors(Current)) {
+        if (isFunctionCall(Current)) {
+          auto *Callee = getFunctionCallCallee(Current);
+          // If not an indirect call, add the node to the CG
+          if (Callee) {
+            auto *Node = BasicBlockNodeMap[Callee];
+            StartNode->addSuccessor(Node);
+          }
+          auto *Next = getFallthrough(Current);
+          if (!Visited.count(Next))
+            Worklist.push_back(Next);
+          continue;
+        }
+
+        if (!GCBI::isPartOfRootDispatcher(Successor)
+            && !Visited.count(Successor))
+          Worklist.push_back(Successor);
+      }
+    } while (!Worklist.empty());
+  }
+
+  for (const auto &[_, Node] : BasicBlockNodeMap) {
+    if (!Node->hasPredecessors())
+      EntryBlocks.emplace_back(Node);
+
+    // Recursive?
+    if (*Node->predecessors().begin() == Node)
+      EntryBlocks.emplace_back(Node);
+  }
+
+  BasicBlockNode FakeNode{ nullptr };
+  auto *RootNode = CG.addNode(FakeNode);
+  CG.setEntryNode(RootNode);
+  for (const auto &Node : EntryBlocks)
+    RootNode->addSuccessor(Node);
+
+  if (StackAnalysisLog.isEnabled()) {
+    std::error_code EC;
+    raw_fd_ostream Out(M.getModuleIdentifier() + "_CG.dot",
+                       EC,
+                       llvm::sys::fs::OF_None);
```

Comment:
Let's not create files in the current directory.
Make an option to represent the output path and, if not empty, throw the graph there (without using `StackAnalysisLog`).